Week 2 Notes covering basic C++ language features

mmcleod89 opened the pull request.

JamieJQuinn commented on 02cpp1/sec02PassByValueOrReference.md:
```diff
@@ -0,0 +1,101 @@
+
+## Using references for output variables
+
+When we use a `return` statement in a function, we are also passing by value, and a copy of the variable that we are returning is made. Just like with inputs to a function, this can be a performance issue if the data that we want to output is large. 
```
I think this should be more precise, i.e. the variable is copied using the object's copy-constructor (I think we can assume students either already know the cop-constructor concept or can look it up).

You should probably also mention return value optimisation as a possible alternative to passing output vars by reference.

mmcleod89 replied:
Agreed. Have added basic discussion of copy constructor and move constructor and how they relate to `return`. Have also changed `vector` to an arbitrary `ImmovableData` class because I realised `vector` _does_ has a move constructor and therefore doesn't have a copy problem when returning.

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
"Stack memory is fast, but limited in size" - fast compared to what? Faster than the heap?

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
Is there possibly a reference to a higher power we can use here? e.g. something from the [core guidelines](https://github.com/isocpp/CppCoreGuidelines)?

mmcleod89 replied:
I have softened this statement a bit to encourage people to think about the dangers of having pointers left empty like this.

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
`updatePtrValue2` isn't defined?

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
`(p)` does not return a bool, it only does so in a statement that expects a bool (e.g. if) where it's implicitly converted. e.g. `if (p) ...` and `bool is_pointing = bool(p)` both work but `bool is_pointing = (p)` does not.

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
```suggestion
The same considerations about using move semantics or references applies when adding unique pointers to containers such as vector. If you want the vector to be the new owner of the memory, then use `std::move`.
```

I wouldn't mention a weird way to do this; `move` works well here.

mmcleod89 replied:
I disagree here; move only works well here if you want to transfer the ownership of the data away from the original unique pointer to the container. But if you just want a container which is keeping track of some things but doesn't own that data (which is not an unusual scenario) then you would need to use a pointer. This is akin to having a pointer to a stack variable in a container, but in this case the variable is a unique ptr (perhaps because it makes more sense for it to be defined on the heap e.g. because it is large).

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
Can there be a summary section here? The pointer material is perfectly detailed IMO but it feels like there are a few key takeaways that could be emphasised.

Could also collate all the good practices you've listed throughout the page

JamieJQuinn commented on 02cpp1/sec03Pointers.md:
```suggestion
- Faster memory access than weak pointers, although unique/shared pointers will match speed of raw pointers if compiled with optimisation.  
```

I don't want to give students any ammunition to claim performance reasons for not using smart pointers.

JamieJQuinn reviewed (changes requested):
Balanced, very well explained and the smart pointer section in particular is carefully done. I've peppered in some suggestions/corrections. Most are suggestions only so take or leave what you feel is right here.

mmcleod89 commented on 02cpp1/sec01StandardLibrary.md:
````diff
@@ -0,0 +1,340 @@
+```
+- Indexing of vectors (and other array like containers) always starts at 0.
+- We can use the `.at(i)` method on a vector to get the value at that position in the vector. If `i` is outside of the range of the vector then this method will throw an error, which can be handled by your program. (See next week when we discuss exceptions.)
+- We can also access a vector using square brackets: `myVector[i]`. This is substantially faster than using `myVector.at(i)`, but does no bounds checking. If `i` is outside of the size of vector then the behaviour is *undefined*: this means the standard does not define what the program should do in this case, and the result will be compiler dependent. *Undefined behaviour should always be avoided ...
````
I see your point; I think it's important to make sure that they understand that these kinds of methods have risk and they shouldn't without thinking about it, but I have softened the wording and made it clear that `[]` is the most common way of accessing vectors in practice

JamieJQuinn replied:
Sure, good enough for me!

mmcleod89 commented on 02cpp1/sec01StandardLibrary.md:
```diff
@@ -0,0 +1,340 @@
+- Indexing of vectors (and other array like containers) always starts at 0.
+- We can use the `.at(i)` method on a vector to get the value at that position in the vector. If `i` is outside of the range of the vector then this method will throw an error, which can be handled by your program. (See next week when we discuss exceptions.)
+- We can also access a vector using square brackets: `myVector[i]`. This is substantially faster than using `myVector.at(i)`, but does no bounds checking. If `i` is outside of the size of vector then the behaviour is *undefined*: this means the standard does not define what the program should do in this case, and the result will be compiler dependent. *Undefined behaviour should always be avoided ...
+- Accessing and modifying vector elements using `[]` has high performance, and can even out-perform traditional C-arrays if compiled with optimisations turned on, which makes `vector` an excellent general purpose class for lists. 
```
This was based on some tests I did on my laptop which did surprise me, but I have changed it to "competitive with" as the results from vectors and C-arrays were quite similar as I recall

JamieJQuinn replied:
Yeah, I think that's ultimately what we want them to takeaway, that we're not losing performance using STL containers

mmcleod89 commented on 02cpp1/sec01StandardLibrary.md:
```diff
@@ -0,0 +1,340 @@
+- Here we have used the `auto` keyword to ask C++ to infer the type for us; this can be useful when we want to avoid writing out lots of types explicitly!
+- This will iterate through all the elements of the list, in order. For each loop iteration, `num` will be initialised with the data for that element in the list. 
+- `num` is initialised the same way as function arguments, so by default the elements of the list are passed by value i.e. a copy of the element is made and stored that in the variable `num`. This means any changes to `num` in the loop code are not reflected in the vector itself. **If you want to avoid copy overheads or make changes to your vector, you should make this variable a reference using  ...
+- Range based loops can be used with other containers which are iterable, like `array` and `map`. 
```
I'm not sure exactly what the key benefit you want to highlight here is?

mmcleod89 commented:
Have addressed points with the exception of adding a summary which to the pointers notes which I don't have time to do right now as it is very late. I think this should be good to merge in though now.
